### DISCUSSION TITLE
Implementation of NewContainer and CreateSnapshot

### DISCUSSION THREAD

**Discussion by tripabhi** [vhive-serverless/vHive#598] [Q&A] [answered]
I am a Master's student at Virginia Tech, affiliated with the Distributed Systems and Storage Laboratory. I am trying to profile the IO behavior (process-level and drive-level) for the different phases defined in vHive's `\ctriface\iface.go` file, e.g., StartVM, PauseVM, CreateSnapshot, LoadSnapshot. To better understand the IO behavior, I am trying to locate the implementation of routines like NewContainer, and CreateSnapshot. I am pretty new to codebases that use RPC; therefore, I am having trouble finding the implementation of the said procedures. For example, in vHive's `\ctriface\iface.go`, we call
```
if _, err := o.fcClient.CreateSnapshot(ctx, req); err != nil && ctxFwd.Err() == context.Canceled{
	logger.WithError(err).Error("failed to create snapshot of the VM")
	return err
}
```

I traced it to [here](https://github.com/ease-lab/firecracker-containerd/blob/load_create_snapshot/firecracker-control/local.go); however, I cannot locate the actual implementation of CreateSnapshot which would make read/write syscalls. Similarly, for NewContainer, I am trying to ascertain the R/W behavior of the process, but I cannot find the implementation details on containerd's [repository](https://github.com/ease-lab/containerd).

I would greatly appreciate help from the vHive community if they could point me to where I can find implementation-specific code, especially those that perform read/write while creating the snapshot files or creating a new container.

**Comment by ustiugov** [comment, marked as answer]
@tripabhi I recommend you to take a look at the full set of modified files  in the [PR](https://github.com/ease-lab/firecracker-containerd/pull/19/files)
